### PR TITLE
Feature/jb 357 incorporate new coco prodict

### PR DESCRIPTION
**Pull request by sei-vturri**
- Integrated coco prodict into the coco_utils, data, and metadata_preprocessor files where applicable
- Removed optional fields from prodict classes in coco_anno.py
- Set optional fields in the schema to {} i.e. Any (tried OneOf to specify multiple possible types... that didn't work)
- Cleaned up test_data.py a bit

### CODE REVIEW

**Comment by sei-nwinski** [comment]
I received the following during the dt2 CPU unit test:

2021-12-15 13:48:24,311 INFO Using category list from     Source: train manifest     Path: models/text_detect/dt2/ut/train/training_manifest.json
2021-12-15 13:48:24,312 INFO Saving predictions in annotation-style format in models/text_detect/dt2/ut/eval/text_detect_val/detections_anno.json
2021-12-15 13:48:24,541 ERROR 'collections.OrderedDict' object has no attribute 'images'
Traceback (most recent call last):
  File "/juneberry/juneberry/scripting.py", line 47, in run_main
    main_fn()
  File "/juneberry/bin/jb_evaluate", line 139, in main
    evaluator.perform_evaluation()
  File "/juneberry/juneberry/evaluation/evaluator.py", line 236, in perform_evaluation
    self.format_evaluation()
  File "/juneberry/juneberry/detectron2/evaluator.py", line 208, in format_evaluation
    coco_utils.save_predictions_as_anno(data_root=self.data_root, dataset_config=str(self.dataset_config.file_path),
  File "/juneberry/juneberry/config/coco_utils.py", line 358, in save_predictions_as_anno
    coco_out = convert_predictions_to_coco(coco_data, predictions, category_list)
  File "/juneberry/juneberry/config/coco_utils.py", line 254, in convert_predictions_to_coco
    'images': coco_data.images,
AttributeError: 'collections.OrderedDict' object has no attribute 'images'
TaskError - taskid:2
Command error: '['jb_evaluate', 'text_detect/dt2/ut', 'data_sets/text_detect_val.json']' returned 255

**Comment by sei-nwinski** [comment]
I think since `coco_utils.convert_jbmeta_to_coco` (line 260) returns a COCO formatted structure, you might be able to use the new COCO prodict to simplify things a bit there.

File: juneberry/config/coco_utils.py [hunk 1]
```diff
@@ -385,7 +386,7 @@ def generate_bbox_images(coco_json: Path, lab, dest_dir: str = None, sample_limi
         logger.info(f"The output directory was not found, so it was created.")
 
     # Load the COCO annotations.
-    coco = jbfs.load_file(coco_json)
+    coco = CocoAnnotations.load(str(coco_json))
```
Comment (sei-nwinski) [inline juneberry/config/coco_utils.py:389]: Now that these are CocoAnnotations, you can do some rewrites of line 399 and beyond, e.g. `image_list = coco["images"]` can become something like `image_list = coco.images` or even eliminated completely. If eliminated, you'd do things like `rand_shuffle(coco.images)` or `for image in coco.images`.

File: juneberry/data.py [hunk 2]
```diff
@@ -404,21 +405,21 @@ def _load_metadata_file(self, filepath: str, new_values: list, remove_image_ids:
...
                 # The categories were changed by the preprocessors. They win out over all others
                 # Categories is a list of id, name
                 # Last preprocessor of all files wins
                 # NOTE: This might not be a good idea to do it every time depending on how the
                 # preprocessors are written, but we'll try this for a while.
                 # TODO: Fix the dichotomy with str vs int labels
-                str_labels = {str(x['id']): x['name'] for x in data['categories']}
-                int_labels = {int(x['id']): x['name'] for x in data['categories']}
+                str_labels = {str(x['id']): x['name'] for x in data.categories}
+                int_labels = {int(x['id']): x['name'] for x in data.categories}
```
Comment (sei-nwinski) [inline juneberry/data.py:422]: Can't these be `x.id` and `x.name`?

Comment (sei-vturri) [reply juneberry/data.py:422]: No I tried that before and it didn't work. I believe this line `data = self._preprocessors(data)` changes the type from prodict to dict so you can't access the field using the dot.

File: juneberry/schemas/coco_anno_schema.json [hunk 3]
```diff
@@ -84,7 +57,8 @@
                         "type": "array",
                         "items": { "type": "number" }
                     },
-                    "iscrowd": { "type": "integer" }
+                    "iscrowd": { "type": "integer" },
+                    "score": { "type": "number" }
                 },
                 "required": [
                     "id",
```
Comment (sei-nwinski) [inline juneberry/schemas/coco_anno_schema.json:64]: I guess this is as close as I can get to leave a line comment, but is there a reason "annotations" isn't required around L74?

Comment (sei-vturri) [reply juneberry/schemas/coco_anno_schema.json:64]: I discussed the required fields for the coco anno file with Andrew and he mentioned use cases where you might not have annotations, just categories and images. I can't remember the specifics. :/

**Review by sei-nwinski — APPROVED** [review]

File: juneberry/schemas/coco_anno_schema.json [hunk 4]
```diff
@@ -19,10 +19,15 @@
             "items": {
                 "type": "object",
                 "properties": {
-                    "id": { "type": "integer" },
-                    "name": { "type": "string" },
-                    "url": { "type": "string" }
+                    "id": { "type": "integer"},
+                    "name": { "type": "integer" },
```
Comment (sei-amellinger) [inline juneberry/schemas/coco_anno_schema.json:23]: Perhaps name and url should be string insetad?

**Review by sei-amellinger — APPROVED** [review]

**Pull request merged**